### DISCUSSION TITLE
Enforcement of ARMI app version with metdata:versions

### DISCUSSION THREAD

**Discussion by drewj-usnctech** [terrapower/armi#1066] [Q&A]
Some reading of #1064 // #1064 has me curious about the `metadata:versions` setting. Digging through the settings implementation, the version from the file is attached as `inputVersion`

https://github.com/terrapower/armi/blob/9d6a405f10a0084b2fea8721f2b6d08c16a0b7a8/armi/settings/settingsIO.py#L222

and `self.liveVersion` is set to `armi.meta.__version__`

But, if they differ, there isn't an explicit error. Any mismatch is noted if there is an invalid setting if an `InvalidSettingsStopProcess` is raised

https://github.com/terrapower/armi/blob/9d6a405f10a0084b2fea8721f2b6d08c16a0b7a8/armi/utils/customExceptions.py#L103-L122

which is created in `SettingsReader._checkInvalidSettings`

https://github.com/terrapower/armi/blob/9d6a405f10a0084b2fea8721f2b6d08c16a0b7a8/armi/settings/settingsIO.py#L226-L244

So long as I don't have any invalid settings in my different ARMI version, then I would never be alerted that my version differs from the one specified in the settings file.

The second question that follows is making this application specific. Is this something that could be picked up by the app version? Our ARMI application has different version numbers from ARMI, so if we wanted to leverage this feature, we would want to tie the version from the input not to the version of the armi library, but to our ARMI application

**Comment by john-science** [comment]
Yeah, the "metadata" version was put there for legal/process reasons. We don't really use or enforce it heavily.  It _is_ interesting that the warning isn't even printed out though.

> Do you think we should use or enforce it strongly?

On occasion, people ask me about the versions of their `Plugin`, `Interface`, or `App`, and if they can track all of those in ARMIs code. While certainly anything is possible, not everyone imports and runs ARMI from an "App" (for instance, if you are using ARMI just to open up a DB from a previous run). But mostly I just think it seems inside-out for ARMI code to do version checking for code it can't see in some other repository. But, perhaps, that's just my gut feeling.

> What do you think of this solution?

Perhaps your `Plugin` / `App` could have it's own bespoke logic for handling versions:

```python
from myapp import __version__

VERSION_SETTING = "myPluginVersion"


class MyPlugin(ArmiPlugin):

    @staticmethod
    @plugins.HOOKIMPL
    def defineSettings():
        return [
            Setting(
                VERSION_SETTING,
                default="uncontrolled",
                label="stuff",
                description="things",
            )
        ]

    @staticmethod
    @plugins.HOOKIMPL
    def defineSettingsValidators(inspector):
        return [
            settingsValidation.Query(
                lambda: inspector.cs[VERSION_SETTING] != __version__,
                "Oh noes! The MyPlugin version doesn't match!",
                "Do you want to force the version to match?",
                lambda: inspector._assignCS(VERSION_SETTING, True),
            )
        ]
```

The above `settingValidation` being in your own repo also gives you a lot of power to handle versions that are meaningful to you. For instance, maybe all you care about is that the major version matches: "4.x", and you could add some validation to allow "4.1.3" to be fine. Or all over your plugins could validate their own versions, to stop your developers from having mix-and-match version issues.

I'm just spit-balling here @drewj-usnctech . Thoughts?

**Reply by drewj-usnctech** [reply]
Should I take away from this that `metadata:version` is a setting that shouldn't be used? I understand needing at least to parse it and put it in input files for QA / record keeping. But, from a user's perspective, it doesn't do what it's advertised to do.

But, to your point, to make ARMI do version enforcement on things that are outside it's scope is not a fair ask. On the assuming y'all have an ARMI application built off ARMI, likely w/ it's own versioning scheme too, would it make sense to allow plugins / apps to define validators for this metadata section? So the settings file could look something like
```yaml
metadata:
  framework version: 0.2.5  # ARMI version
  custom app version: x.y.z
```
and then it's the responsibility of the custom app to validate `x.y.z`

Version validation is finicky because, like you pointed out, what if we need to force a looser version `4.x` vs. a more explicit version `4.9.1`? I would say that it depends on the intent behind putting versions in a settings file in the first place. If it really is intended to enforce that a case cannot be ran without a specific version, then literal matching would suffice in my book.

In the spirit of ARMI plugins, I really like the idea of allowing apps / plugins to do their own validation. More trying to nail down where the user is expected to put this. Using your example, if someone read the docs, they could have a settings file like
```yaml
metadata:
  version: uncontrolled  # to avoid ARMI errors
settings:
  myPluginVersion: 4.9.1
```
which feels weird? If someone was reading this without the knowledge of what `metadata:versions` and `settings:myPluginVersions` meant, what should they take away?

**Reply by john-science** [reply]
Give me a hot minute to think about this.

_IF_ we're okay with enfocing the `1.2.3` versioning for ARMI, and related plugins/apps, then I could write some helper functions that could easily tell if `4.1.3` meets the requirement `4.1`, and similar.

Then we could have any plugin add a new setting (say to the metadata section) and a validator that just calls my function.

That would mean that ARMI would contain the code to do what you want, but ARMI wouldn't be responsible for understanding outside plugins, and it would be the plugin developer who would be responsible.  I think that would meet your needs.

Our current solution to this problem, I suppose, is that the ARMI `App` we create points to ultra-specific commits to each plugin (in its own repo), including a repo for data files. And we have CI to test that all these versions/commits work together. So I believe the trustworthiness of that process is WHY we don't have this feature you suggest all flushed out.  But maybe just adding a couple of helper methods for you would be low-effort enough I might as well do it.

Thoughts?

**Reply by drewj-usnctech** [reply]
I'm very much on board for

> Then we could have any plugin add a new setting (say to the metadata section) and a validator that just calls my function

> That would mean that ARMI would contain the code to do what you want, but ARMI wouldn't be responsible for understanding outside plugins, and it would be the plugin developer who would be responsible

we could then have our "primary" plugin do the validation on our app version, ARMI validates ARMI's version, and any other plugins (hosted outside our "core" app repo) could do their own version validation

**Reply by john-science** [reply]
Okay, this seems like a light-weight solution to me.  

I opened [a ticket](https://github.com/terrapower/armi/issues/1095) to build said helper function.
